File: Problems/Algorithms/695. Max Area of Island/MaxArea.kt
class Solution {
    fun maxAreaOfIsland(grid: Array<IntArray>): Int {
        val n = grid.size
        val m = grid[0].size
        
        val visited = Array(n) { IntArray(m) { 0 } }
        
        var ans = 0
        for (r in 0..n-1) {
            for (c in 0..m-1) {
                ans = maxOf(ans, dfs(grid, visited, n, m, r, c))
            }
        }
        
        return ans
    }
    
    private fun isValid(n: Int, m: Int, r: Int, c: Int): Boolean {
        return r >= 0 && r < n && c >= 0 && c < m
    }
    
    private fun dfs(grid: Array<IntArray>, visited: Array<IntArray>, n: Int, m: Int, r: Int, c: Int): Int {
        if (!isValid(n, m, r, c) || visited[r][c] == 1 || grid[r][c] == 0) return 0
        
        visited[r][c] = 1
        return 1 + dfs(grid, visited, n, m, r+1, c) + dfs(grid, visited, n, m, r-1, c) + dfs(grid, visited, n, m, r, c+1) + dfs(grid, visited, n, m, r, c-1)
    }
}

File: Problems/Algorithms/695. Max Area of Island/max_area.go
func maxAreaOfIsland(grid [][]int) int {
    n, m := len(grid), len(grid[0])

    visited := make([][]int, n)
    for row := range visited {
        visited[row] = make([]int, m)
    }
    neighbors := [][]int{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}

    best := 0
    for r := 0; r < n; r++ {
        for c := 0; c < m; c++ {
            best = max(best, dfs(grid, visited, neighbors, n, m, r, c))
        }
    }

    return best
}

func dfs(grid [][]int, visited [][]int, neighbors [][]int, n int, m int, r int, c int) int {
    if !isValid(n, m, r, c) || visited[r][c] == 1 || grid[r][c] == 0 {
        return 0
    }

    visited[r][c] = 1

    ans := 1
    for _, neighbor := range neighbors {
        ans += dfs(grid, visited, neighbors, n, m, r+neighbor[0], c+neighbor[1])
    }

    return ans
}

func isValid(n int, m int, r int, c int) bool {
    return r >= 0 && r < n && c >= 0 && c < m
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}
